python/ray/dag/collective_node.py

```diff
@@ -0,0 +1,197 @@
+
+
+class CollectiveGroup:
+    """Represent metadata for a NCCL collective method."""
+
+    def __init__(
+        self,
+        input_nodes: List[DAGNode],
+        op: CollectiveOp,
+        transport: Union[str, GPUCommunicator] = TorchTensorType.NCCL,
+    ):
+        self._input_nodes: List[DAGNode] = input_nodes
+        if len(self._input_nodes) == 0:
+            raise ValueError("Expected input nodes for a collective group")
+        if len(set(self._input_nodes)) != len(self._input_nodes):
+            raise ValueError("Expected unique input nodes for a collective group")
+
+        self._actor_handles: List["ray.actor.ActorHandle"] = []
+        for input_node in self._input_nodes:
+            actor_handle = input_node._get_actor_handle()
+            assert actor_handle is not None, "Expected a actor handle"
+            self._actor_handles.append(actor_handle)
+        if len(set(self._actor_handles)) != len(self._actor_handles):
+            raise ValueError("Expected unique actor handles for a collective group")
+
+        self._op = op
+        assert isinstance(
+            self._op, ReduceOp
+        ), "Other collective ops are not implemented"
+        self._type_hint = TorchTensorType(transport=transport, _direct_return=True)
+        if isinstance(transport, GPUCommunicator):
+            assert set(transport.get_actor_handles()) == set(
+                self._actor_handles
+            ), "Expected actor handles to match the custom NCCL group"
+
+    def __str__(self) -> str:
+        return (
+            f"CollectiveGroup("
+            f"_input_nodes={self._input_nodes}, "
+            f"_actor_handles={self._actor_handles}, "
+            f"_op={self._op}, "
+            f"_type_hint={self._type_hint})"
+        )
+
+    @property
+    def actor_handles(self) -> List["ray.actor.ActorHandle"]:
+        return self._actor_handles
+
+    @property
+    def type_hint(self) -> TorchTensorType:
+        return self._type_hint
+
+    def init_nccl_group(self) -> str:
+        type_hint = self._type_hint
+        if type_hint.nccl_group_id is not None:
+            # The NCCL group has already been initialized.
+            return type_hint.nccl_group_id
+        nccl_group_id = _init_nccl_group(
+            self._actor_handles, type_hint.get_custom_nccl_group()
+        )
+        type_hint.set_nccl_group_id(nccl_group_id)
+        return nccl_group_id
+
+    def get_nccl_group(self) -> GPUCommunicator:
+        if self._type_hint.nccl_group_id is not None:
+            ctx = ChannelContext.get_current()
+            nccl_group = ctx.nccl_groups[self._type_hint.nccl_group_id]
+        elif self._type_hint.get_custom_nccl_group() is not None:
+            nccl_group = self._type_hint.get_custom_nccl_group()
+        else:
+            raise ValueError("Expected a NCCL group")
+        return nccl_group
+
+    def method(self, tensor: "torch.Tensor"):
```

Comment:
docstring and annotate return type

Should we call it `apply_op` or something along those lines?

Reply:
Ack.